### PR TITLE
rename `<stdatomic.h>` to `<__msvc_cxx_stdatomic.h>`

### PR DESCRIPTION
**Pull request by barcharcraz**
we are adding C11 atomics support to the compiler/vcruntime. because of this `<stdatomic.h>` needs to become part of vcruntime. This PR renames the STL's "compatibility" stdatomic.h to `__msvc_cxx_stdatomic.h` and updates references to `stdatomic.h` to instead point to `__msvc_cxx_stdatomic.h`. The "real" stdatomic.h is added to vcruntime in an internal mirror of this PR: MSVC-PR-423692, and it simply includes either `__msvc_cxx_stdatomic.h` or `vcruntime_c11_stdatomic.h` depending on `__cplusplus`.

The one inclusion of "stdatomic.h" in our tests was updated, so our tests should continue to pass, even if the runtime used to build the STL does not have `stdatomic.h`.

### CODE REVIEW

File: stl/inc/__msvc_all_public_headers.hpp [hunk 1]
```diff
@@ -139,13 +139,11 @@
 #include <vector>
 
 #ifndef _M_CEE_PURE
+#include <__msvc_cxx_stdatomic.hpp>
```
Comment (CaseyCarter) [inline stl/inc/__msvc_all_public_headers.hpp:142]: This isn't a public header, per se. Should we TRANSITION and switch back to including `stdatomic.h` once things stabilize? (Also in `header-units.json` and `P0943R6_stdatomic_h``.)

File: stl/inc/header-units.json [hunk 2]
```diff
@@ -6,6 +6,7 @@
     "BuildAsHeaderUnits": [
         // "__msvc_all_public_headers.hpp", // for testing, not production
         "__msvc_chrono.hpp",
+        "__msvc_cxx_stdatomic.hpp",
```
Comment (CaseyCarter) [inline stl/inc/header-units.json:9]: Don't we need to remove `stdatomic.h` from this list as well?

File: stl/CMakeLists.txt [hunk 3]
```diff
@@ -4,6 +4,7 @@
 set(HEADERS
```
Comment (CaseyCarter) [inline stl/CMakeLists.txt:4]: Do we also need updates to `stl/modules/std.ixx`, `tests/std/tests/include_each_header_alone_matrix.lst`, `tests/std/tests/P1502R1_standard_library_header_units/importable_cxx_library_headers.jsonc`, `tests/std/tests/P1502R1_standard_library_header_units/test.cpp`, and `tests/std/include/test_header_units_and_modules.hpp`?

File: stl/inc/__msvc_all_public_headers.hpp (hunk 1 above)
Comment (barcharcraz) [reply stl/inc/__msvc_all_public_headers.hpp:142]: well, header-units lists many non-public headers. stdatomic_h 🤷 We could transition it back, but it's reasonable to just include the one that's part of "us" so to speak

File: stl/inc/header-units.json (hunk 2 above)
Comment (barcharcraz) [reply stl/inc/header-units.json:9]: I don't _think_ so, in that `<stdatomic.h>` should, in fact, still be a header unit.

Having another header-units.json file in vcruntime would be problematic.

File: stl/CMakeLists.txt (hunk 3 above)
Comment (barcharcraz) [reply stl/CMakeLists.txt:4]: * `std.ixx`: no, it doesn't include stdatomic (as c++ stdatomic only makes global aliases for std:: stuff
* `tests/std/tests/P1502R1_standard_library_header_units/importable_cxx_library_headers.jsonc` ditto
* `include_each_header_alone_matrix.lst`: maybe, but really it should be testing the public header.
* `tests/std/tests/P1502R1_standard_library_header_units/test.cpp`: no, does not import anything stdatomic related, just normal atomic
* `tests/std/include/test_header_units_and_modules.hpp`: ditto

File: stl/inc/header-units.json (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/header-units.json:9]: Won't the header-unit tests fail on GitHub until we update our toolset compiler to one that includes `stdatomic.h` in VCRuntime?

Comment (CaseyCarter) [reply stl/inc/header-units.json:9]: ...or will it succeed by using the toolset compiler's `stdatomic.h`?

Comment (barcharcraz) [reply stl/inc/header-units.json:9]: I thought the latter, but apparently the former. It has been fixed

**Review by CaseyCarter — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by frederick-vs-ja** [comment]
I think this error message will be confusing after the changes. It still says `<stdatomic.h>`, but if one include `<stdatomic.h>` in C11/17 mode (after the corresponding changes in vcruntime), this `#error` won't be encountered.

Perhaps we should say something like `<__msvc_cxx_stdatomic.hpp> is an internal header. It is incompatible with C and should not be directly included.`.

https://github.com/microsoft/STL/blob/996ad1f4cbfcb73610a8946d846cdb9d809fd5e3/stl/inc/__msvc_cxx_stdatomic.hpp#L13-L16

**Pull request merged**